File: apps/desktop/shell/src-tauri/src/notifications.rs
//! Notifications: the OS notification channel through
//! `tauri-plugin-notification` (R08).
//!
//! The truth law: delivery is NEVER fabricated — permission-denied before
//! the grant, invalid-request for malformed notifications, and `delivered`
//! only after the OS accepted the handoff.

use tauri_plugin_notification::NotificationExt;

use tauri::AppHandle;

use crate::ipc::{CommandResult, ShellError, ShellNotification, ShellNotifyOutcome, ShellPermissionState};

pub struct Notifications;

impl Notifications {
    /// The truthful current permission state.
    pub fn permission(app: &AppHandle) -> CommandResult<ShellPermissionState> {
        let state = app.notification().permission_state();
        let (granted, can_request) = match state {
            tauri_plugin_notification::PermissionState::Granted => (true, false),
            tauri_plugin_notification::PermissionState::Denied => {
                // A denied OS state may be soft (askable) or hard (settings);
                // request_permission is the honest probe-free resolution —
                // canRequest stays true until a request answers denied.
                (false, true)
            }
            tauri_plugin_notification::PermissionState::Prompt => (false, true),
            tauri_plugin_notification::PermissionState::Default => (false, true),
            _ => (false, true),
        };
        Ok(ShellPermissionState {
            granted,
            can_request: can_request && !granted,
            reason: if granted { None } else { Some("the OS notification permission is not granted".into()) },
        })
    }

    /// Ask the user where possible; answers the resulting state.
    pub async fn request_permission(app: AppHandle) -> CommandResult<ShellPermissionState> {
        let result = app.notification().request_permission().await;
        match result {
            Ok(tauri_plugin_notification::PermissionState::Granted) => {
                Ok(ShellPermissionState { granted: true, can_request: false, reason: None })
            }
            Ok(_) => Ok(ShellPermissionState {
                granted: false,
                can_request: false,
                reason: Some("the user or OS denied the notification permission".into()),
            }),
            Err(e) => Err(ShellError::new("unavailable", format!("requesting notification permission failed: {e}"))),
        }
    }

    /// Show one notification; the outcome is typed and honest.
    pub fn show(app: &AppHandle, notification: ShellNotification) -> CommandResult<ShellNotifyOutcome> {
        if notification.title.trim().is_empty() {
            return Ok(ShellNotifyOutcome::NotDelivered {
                reason: "invalid-request",
                detail: format!("notification.title: expected a non-empty string, got '{}'", notification.title),
            });
        }
        let permission = Self::permission(app)?;
        if !permission.granted {
            return Ok(ShellNotifyOutcome::NotDelivered {
                reason: "permission-denied",
                detail: "the OS notification permission is not granted".into(),
            });
        }
        let mut builder = app.notification().builder().title(&notification.title);
        if let Some(body) = &notification.body {
            builder = builder.body(body);
        }
        // The real OS handoff (the plugin routes to NSUserNotification /
        // the Windows notification channel / libnotify).
        builder
            .show()
            .map_err(|e| ShellError::new("unavailable", format!("the OS notification channel failed: {e}")))?;
        Ok(ShellNotifyOutcome::Delivered)
    }
}

File: apps/web/src/platform/notifications.ts
/**
 * @wfx/app-web — the Web NotificationPort (R07).
 *
 * The Web adapter's notification capability over the REAL Web Notifications
 * API — permission-gated, never fabricating delivery:
 *
 * - `permission()` reports the truthful current state (`granted` /
 *   `denied` / `default`) with `canRequest` true exactly while the state is
 *   `default` (a hard `denied` cannot be re-asked in most browsers without
 *   the user changing site settings — the honest reading).
 * - `requestPermission()` asks the USER through the browser's real prompt.
 *   It never auto-grants and never assumes; the answer is whatever the
 *   browser reports.
 * - `notify()` constructs one REAL notification only when permission is
 *   granted. A denied permission answers the typed
 *   `{ delivered: false, reason: "permission-denied" }`; a malformed
 *   request answers `invalid-request`; a construction failure answers
 *   `unavailable`. `delivered: true` means the browser accepted the
 *   notification for display — nothing more is claimed (read receipts and
 *   display guarantees belong to the OS, not to this port).
 *
 * The port is CONSTRUCTED ONLY when the environment has the Notification
 * API (see `capabilities.ts` — the bundle declares `notifications: true`
 * exactly then). A boot context without the API declares the capability
 * absent and provides no port: the honest unsupported state, never a stub
 * that pretends.
 */

import type {
  NotificationOutcome,
  NotificationPermission,
  NotificationPort,
  NotificationRequest,
} from "@wfx/platform-contracts";

import type { NotificationPermissionState, WebEnvironment } from "./environment";

/** Options for {@link createWebNotificationPort}. */
export interface WebNotificationPortOptions {
  /** The environment (must carry the Notification API — the caller checks). */
  readonly environment: WebEnvironment;
}

/** Create the permission-gated Web notification port. */
export function createWebNotificationPort(options: WebNotificationPortOptions): NotificationPort {
  const api = options.environment.notificationApi;
  const ctor = options.environment.notificationCtor;
  if (api === null || ctor === null) {
    // Defensive: the capability bundle only constructs this port when the
    // API exists. If reached without it, every operation answers the
    // honest unavailable outcome — never a fabricated delivery.
    return {
      permission: async () => ({
        granted: false,
        canRequest: false,
        reason: "the Web Notifications API is not available in this context",
      }),
      requestPermission: async () => ({
        granted: false,
        canRequest: false,
        reason: "the Web Notifications API is not available in this context",
      }),
      notify: async () => ({
        delivered: false,
        reason: "unavailable",
        detail: "the Web Notifications API is not available in this context",
      }),
    };
  }

  function permissionFrom(state: NotificationPermissionState): NotificationPermission {
    switch (state) {
      case "granted":
        return { granted: true, canRequest: false };
      case "denied":
        return {
          granted: false,
          canRequest: false,
          reason: "notification permission was denied — the browser requires the user to re-enable it in site settings",
        };
      case "default":
        return {
          granted: false,
          canRequest: true,
          reason: "notification permission has not been asked for yet",
        };
    }
  }

  return {
    async permission(): Promise<NotificationPermission> {
      return permissionFrom(api.permission);
    },

    async requestPermission(): Promise<NotificationPermission> {
      try {
        const state = await api.requestPermission();
        return permissionFrom(state);
      } catch (thrown) {
        // Some browsers throw when the prompt cannot be shown (e.g. denied
        // prompts re-asked, or no user gesture). The honest answer is the
        // CURRENT state — never an assumed grant.
        return {
          granted: false,
          canRequest: api.permission === "default",
          reason: `the permission prompt could not be shown: ${
            thrown instanceof Error ? thrown.message : String(thrown)
          }`,
        };
      }
    },

    async notify(request: NotificationRequest): Promise<NotificationOutcome> {
      if (typeof request?.title !== "string" || request.title.trim().length === 0) {
        return {
          delivered: false,
          reason: "invalid-request",
          detail: "title: expected a non-empty string",
        };
      }
      if (api.permission !== "granted") {
        return {
          delivered: false,
          reason: "permission-denied",
          detail: `the current permission state is '${api.permission}' — only a granted permission can show notifications`,
        };
      }
      try {
        // The REAL handoff: one OS/browser notification. Constructing it IS
        // the delivery claim; nothing beyond acceptance is claimed.
        new ctor(request.title, {
          ...(request.body !== undefined ? { body: request.body } : {}),
          ...(request.itemId !== undefined ? { tag: `wfx:${request.category}:${request.itemId}` } : {}),
        });
        return { delivered: true };
      } catch (thrown) {
        return {
          delivered: false,
          reason: "unavailable",
          detail: `the notification could not be constructed: ${
            thrown instanceof Error ? thrown.message : String(thrown)
          }`,
        };
      }
    },
  };
}
